### PR TITLE
Support containerd for network chaos

### LINKED ISSUE #161: Support containerd when getting networknamespace
## Feature Request

As we have use `kind` as suggested development environment in development guide #160, and `minikube` linux mirrors doesn't support netem (because kernel doesn't have `sch_netem` module), we should support `containerd` runtime.

Now we use docker pkg to get pid by container id, and we can do the same thing with [`containerd`](https://godoc.org/github.com/containerd/containerd). We can filter the container out from  `client.Containers()` and find pid with `container.Task().Pid()`

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

### What problem does this PR solve? 
fixes #161 

### What is changed and how does it work?
Add a `--runtime` flag to chaos-daemon to specify different runtime.
Add support for using containerd.

### Check List 

Tests 

 - Unit test
 - E2E test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has Go code change
 - Has CI related scripts change
 - Has Terraform scripts change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: cmd/chaos-daemon/main.go [hunk 1]
```diff
@@ -28,13 +28,15 @@ import (
...
-	rawPort      string
+	printVersion     bool
+	rawPort          string
+	containerRuntime string
 )
 
 func init() {
 	flag.BoolVar(&printVersion, "version", false, "print version information and exit")
 	flag.StringVar(&rawPort, "port", "", "the port which server listens on")
+	flag.StringVar(&containerRuntime, "runtime", "docker", "")
```
Comment (yeya24) [inline cmd/chaos-daemon/main.go]: Forget to add the help message...

**Comment by yeya24** [comment]
> Have you test both in containerd and docker? @yeya24

Yes, I have tested in containerd environment. You can use kind cluster to test it with containerd because kind use containerd as the default CRI.

**Comment by YangKeao** [comment]
Good job :rocket:! 

Config field should also be added in helm charts config (in `/helm/chaos-mesh`).

File: pkg/chaosdaemon/server.go [hunk 2]
```diff
@@ -45,22 +45,26 @@ func newServer() (*Server, error) {
...
 		log.Error(err, "failed to create server")
+		return
 	}
 	pb.RegisterChaosDaemonServer(s, chaosDaemonServer)
 
 	reflection.Register(s)
 
 	if err := s.Serve(lis); err != nil {
 		log.Error(err, "failed to serve")
+		return
```
Comment (yeya24) [inline pkg/chaosdaemon/server.go]: Can I ask why we don't add return when there is an error in this function? Is it a bug?

Comment (zhouqiang-cl) [reply pkg/chaosdaemon/server.go]: @YangKeao @cwen0 😄

Comment (zhouqiang-cl) [reply pkg/chaosdaemon/server.go]: @yeya24 Because of The Spring Festival, maybe we did not response very quickly.

Comment (you06) [reply pkg/chaosdaemon/server.go]: It's strange don't return when meet error. But if we don't return anything just stop this function, the process will exit normally. What about return an error in this function?

Comment (zhouqiang-cl) [reply pkg/chaosdaemon/server.go]: I think here we should add return, I think it is a bug.

Comment (cwen0) [reply pkg/chaosdaemon/server.go]: Yes, this is a bug, I think we should return the error and handle this error in main function.

**Comment by zhouqiang-cl** [comment]
@YangKeao @you06 @mahjonp @cwen0 can any of your guys test this, thank you

File: helm/chaos-mesh/values.yaml [hunk 3]
```diff
@@ -48,6 +48,13 @@ chaosDaemon:
   imagePullPolicy: Always
   port: 31767
 
+  # runtime specifies which container runtime to use. Currently
+  # we only supports docker and containerd.
+  runtime: docker
+
+  # socketPath specifies the container runtime socket.
+  socketPath: /var/run/docker.sock
```
Comment (cwen0) [inline helm/chaos-mesh/values.yaml]: how to set it if the runtime is containerd?

File: cmd/chaos-daemon/main.go [hunk 4]
```diff
@@ -62,5 +64,5 @@ func main() {
 		os.Exit(1)
 	}
 	log.Info("starting server")
-	chaosdaemon.StartServer("0.0.0.0", port)
+	chaosdaemon.StartServer("0.0.0.0", port, containerRuntime)
```
Comment (cwen0) [inline cmd/chaos-daemon/main.go]: add more log for `port` and `containerRuntime`?

**Comment by cwen0** [comment]
> thank

test in kind cluster 

* netem delay 
```
64 bytes from 10.244.4.8: seq=26 ttl=62 time=138.101 ms
64 bytes from 10.244.4.8: seq=27 ttl=62 time=127.035 ms
64 bytes from 10.244.4.8: seq=28 ttl=62 time=102.897 ms
64 bytes from 10.244.4.8: seq=29 ttl=62 time=139.161 ms
64 bytes from 10.244.4.8: seq=30 ttl=62 time=118.765 ms
64 bytes from 10.244.4.8: seq=31 ttl=62 time=138.217 ms
64 bytes from 10.244.4.8: seq=32 ttl=62 time=130.955 ms
64 bytes from 10.244.4.8: seq=33 ttl=62 time=0.135 ms
64 bytes from 10.244.4.8: seq=34 ttl=62 time=0.188 ms
64 bytes from 10.244.4.8: seq=35 ttl=62 time=0.165 ms
64 bytes from 10.244.4.8: seq=36 ttl=62 time=0.186 ms
64 bytes from 10.244.4.8: seq=37 ttl=62 time=0.119 ms
64 bytes from 10.244.4.8: seq=38 ttl=62 time=0.141 ms
``` 

* network partion 

```
/ # ping 10.244.5.9
PING 10.244.5.9 (10.244.5.9): 56 data bytes
ping: sendto: Operation not permitted
/ # ping 10.244.5.9
PING 10.244.5.9 (10.244.5.9): 56 data bytes
64 bytes from 10.244.5.9: seq=0 ttl=62 time=0.331 ms
64 bytes from 10.244.5.9: seq=1 ttl=62 time=0.191 ms
64 bytes from 10.244.5.9: seq=2 ttl=62 time=0.209 ms
``` 

I have tested it in the Kind cluster and the network chaos works fine.

**Comment by cwen0** [comment]
> @YangKeao @you06 @mahjonp @cwen0 can any one of your guys test this, thank you


test in the kind cluster 

* netem delay 
```
64 bytes from 10.244.4.8: seq=26 ttl=62 time=138.101 ms
64 bytes from 10.244.4.8: seq=27 ttl=62 time=127.035 ms
64 bytes from 10.244.4.8: seq=28 ttl=62 time=102.897 ms
64 bytes from 10.244.4.8: seq=29 ttl=62 time=139.161 ms
64 bytes from 10.244.4.8: seq=30 ttl=62 time=118.765 ms
64 bytes from 10.244.4.8: seq=31 ttl=62 time=138.217 ms
64 bytes from 10.244.4.8: seq=32 ttl=62 time=130.955 ms
64 bytes from 10.244.4.8: seq=33 ttl=62 time=0.135 ms
64 bytes from 10.244.4.8: seq=34 ttl=62 time=0.188 ms
64 bytes from 10.244.4.8: seq=35 ttl=62 time=0.165 ms
64 bytes from 10.244.4.8: seq=36 ttl=62 time=0.186 ms
64 bytes from 10.244.4.8: seq=37 ttl=62 time=0.119 ms
64 bytes from 10.244.4.8: seq=38 ttl=62 time=0.141 ms
``` 

* network partion 

```
/ # ping 10.244.5.9
PING 10.244.5.9 (10.244.5.9): 56 data bytes
ping: sendto: Operation not permitted
/ # ping 10.244.5.9
PING 10.244.5.9 (10.244.5.9): 56 data bytes
64 bytes from 10.244.5.9: seq=0 ttl=62 time=0.331 ms
64 bytes from 10.244.5.9: seq=1 ttl=62 time=0.191 ms
64 bytes from 10.244.5.9: seq=2 ttl=62 time=0.209 ms
``` 

I have tested it in the Kind cluster and the network chaos works fine.

**Comment by zhouqiang-cl** [comment]
> > @YangKeao @you06 @mahjonp @cwen0 can any one of your guys test this, thank you
> 
> test in the kind cluster
> 
> * netem delay
> 
> ```
> 64 bytes from 10.244.4.8: seq=26 ttl=62 time=138.101 ms
> 64 bytes from 10.244.4.8: seq=27 ttl=62 time=127.035 ms
> 64 bytes from 10.244.4.8: seq=28 ttl=62 time=102.897 ms
> 64 bytes from 10.244.4.8: seq=29 ttl=62 time=139.161 ms
> 64 bytes from 10.244.4.8: seq=30 ttl=62 time=118.765 ms
> 64 bytes from 10.244.4.8: seq=31 ttl=62 time=138.217 ms
> 64 bytes from 10.244.4.8: seq=32 ttl=62 time=130.955 ms
> 64 bytes from 10.244.4.8: seq=33 ttl=62 time=0.135 ms
> 64 bytes from 10.244.4.8: seq=34 ttl=62 time=0.188 ms
> 64 bytes from 10.244.4.8: seq=35 ttl=62 time=0.165 ms
> 64 bytes from 10.244.4.8: seq=36 ttl=62 time=0.186 ms
> 64 bytes from 10.244.4.8: seq=37 ttl=62 time=0.119 ms
> 64 bytes from 10.244.4.8: seq=38 ttl=62 time=0.141 ms
> ```
> 
> * network partion
> 
> ```
> / # ping 10.244.5.9
> PING 10.244.5.9 (10.244.5.9): 56 data bytes
> ping: sendto: Operation not permitted
> / # ping 10.244.5.9
> PING 10.244.5.9 (10.244.5.9): 56 data bytes
> 64 bytes from 10.244.5.9: seq=0 ttl=62 time=0.331 ms
> 64 bytes from 10.244.5.9: seq=1 ttl=62 time=0.191 ms
> 64 bytes from 10.244.5.9: seq=2 ttl=62 time=0.209 ms
> ```
> 
> I have tested it in the Kind cluster and the network chaos works fine.

great. Maybe we should also modify readme

**Comment by yeya24** [comment]
@cwen0  Hello, I have addressed all of your comments. PTAL

File: pkg/chaosdaemon/server.go [hunk 5]
```diff
@@ -15,6 +15,7 @@ package chaosdaemon
 
 import (
 	"fmt"
+	"github.com/pkg/errors"
```
Comment (cwen0) [inline pkg/chaosdaemon/server.go]: format this

File: pkg/chaosdaemon/server.go [hunk 6]
```diff
@@ -45,22 +46,26 @@ func newServer() (*Server, error) {
 	}, nil
 }
 
-func StartServer(host string, port int) {
+// StartServer starts chaos-daemon.
+func StartServer(host string, port int, containerRuntime string) error {
 	lis, err := net.Listen("tcp", fmt.Sprintf("%s:%d", host, port))
 	if err != nil {
-		log.Error(err, "failed to listen")
+		return errors.Wrap(err, "failed to listen")
```
Comment (cwen0) [inline pkg/chaosdaemon/server.go]: Don't use `errors.Wrap` in this pr first, you can push a single PR later to unify all error handling.

Comment (yeya24) [reply pkg/chaosdaemon/server.go]: Do you mean that I just need to log and return that error directly?

Comment (yeya24) [reply pkg/chaosdaemon/server.go]: Done

**Comment by zhouqiang-cl** [comment]
@cwen0 ptal again

**Pull request merged**